### DISCUSSION TITLE
Embed xml file to a pdf

### DISCUSSION THREAD

**Discussion by HelenaBadaoui** [prawnpdf/prawn#1231] [Q&A]
Hello guys, 

I need to attach a XML file within a PDF file to be PDF/A3 compliant.
I am using rails and so far I had no issues to generate my PDF (using Prawn gem) and my XML separately but I do not know how to "combine" them.

Any ideas ?

Cheers

**Comment by gettalong** [comment]
Tl;dr: Prawn doesn't support it out of the box but you can do it manually.

See https://github.com/prawnpdf/prawn/issues/1053, https://github.com/prawnpdf/prawn/issues/1063, https://github.com/prawnpdf/prawn/issues/877 and https://gist.github.com/fidothe/32613b44a5b6588531d9.

There are probably some other issues/messages in the old prawn Google group regarding file embedding.

**Reply by HelenaBadaoui** [reply]
Thank you for your response :) 
I'm struggling a lot with this issue, even though I tried understanding the following link https://gist.github.com/fidothe/32613b44a5b6588531d9
As you said, I can do it manually, do you know how ? I've been on this issue for the past week and I have no idea how to do it.
Thanks in advance

**Comment by SarahOuf** [comment]
Hello @HelenaBadaoui, were you able to find a solution to this issue? As I am stuck in the same problem.
Thank you in advance!

**Comment by reiz** [comment]
I have the same issue. A solution would be highly appreciated :-)

**Comment by timokleemann** [comment]
@HelenaBadaoui, @SarahOuf and @reiz, you might want to have a look at the [prawn-attachment gem](https://github.com/invopop/prawn-attachment).

Inside your Prawn class you can then do something like this:

```
require "prawn"
require "prawn/attachment"

Prawn::Document.generate("hello.pdf") do
  text "Some PDF content"
  Tempfile.create do |f|
    f.write("Some XML content")
    f.rewind
    attach("data.xml", f.read) # <-- This method is provided by the prawn-attachment gem
  end
end
```

Use something like [Adobe Acrobat Reader](https://get.adobe.com/de/reader/) then to open the PDF and verify the XML attachment is there. (Unfortunately, Apple's *Preview* can't show PDF attachments if you're a Mac user.)

**Reply by tvdeyen** [reply]
`prawn-attachment` looks promising, but for full E-Rechnung compatibility there seems to be still something missing https://github.com/invopop/prawn-attachment/pull/5

**Reply by timokleemann** [reply]
We managed to set up **full ZUGfERD / XFactur / XRechnung compatibility** in our [invoicing app](https://www.billingengine.com) based on Ruby on Rails. Feel free to check it out. All invoice PDFs generated with it are fully compliant.
